### PR TITLE
Feat: add new doc for search best practice

### LINKED ISSUE #115: Suggested doc: best practices for site search
There are several tools available for projects to use to let their readers search onsite documentation.
* what are the options
* what are the pros/cons of each
* when is it best to use one over another

### PR DESCRIPTION
**Pull request by thisisobate**
Fixes #115

### CODE REVIEW

File: docs/search-best-practice.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Best Practices for Documentation Search
+Documentation (docs) search is an important part of developer experience. It makes life so easy for developers by enabling them narrow down to sections they’re looking for in the documentation. Choosing the right tool for docs search can be a hassle. 
+
+Here at CNCF,we’ve explored different search tools and we’ve decided to narrow our options to this 3:
+
+1. Docsearch by Algolia
+1. Google search
+1. Lunr
+
```
Comment (nate-double-u) [inline docs/search-best-practice.md:9]: Could we invert this list? Start with Lunr, end with Algolia?

File: docs/search-best-practice.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+### Pros
+- Easy to configure and setup
+- Multi-language support
+- Support for image search
+- Search across a specified collection of sites or pages
+- No daily limits for query or records.
+
+
+### Cons
```
Comment (nate-double-u) [inline docs/search-best-practice.md:40]: I'd need to find some data around this, but i think that google search may have some issues around being used in various countries.

File: docs/search-best-practice.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+### Pros
+- Support for offline search
+- No external package dependency
+- Completely free and open source
+- Full power for customization
+- Search index is completely managed and hosted by the owner.
+
+
+### Cons
+- Difficult to configure and setup
```
Comment (nate-double-u) [inline docs/search-best-practice.md:59]: *can* be difficult to configure and setup. If a team is already using hugo/docsy, this should be _very_ easy to setup.

File: docs/search-best-practice.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+- Requires javascript knowledge
+
+# When Is It Best To Use One Over Another?
+If you are looking to create a search capability for your open source project without having to depend on a 3rd party service, then you should consider using [Lunr](https://lunrjs.com/). You can take a look at [this](https://github.com/vitessio/website/pull/1119) pull request to see how we used Lunr to implement search in Vitess docs website.  
+
+If you are looking for a quick and easy way to setup search without caring too much about cost, customization, and 3rd party dependency, then you should consider going for [Docsearch by Algolia](https://docsearch.algolia.com/).
```
Comment (nate-double-u) [inline docs/search-best-practice.md:65]: I disagree -- everytime I've seen someone try to setup Algolia it's gone wrong. I think it's one of the most complicated to set up

**Review by nate-double-u — COMMENTED** [review]
Hi @thisisobate, thanks for this!

I've made a couple of comments inline, but this is a great start!

File: docs/search-best-practice.md (hunk 3 above)
Comment (thisisobate) [reply docs/search-best-practice.md:59]: oh true.

File: docs/search-best-practice.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
+# Best Practices for Documentation Search
+Documentation (docs) search is an important part of developer experience. It makes life so easy for developers by enabling them narrow down to sections they’re looking for in the documentation. Choosing the right tool for docs search can be a hassle. 
+
+Here at CNCF,we’ve explored different search tools and we’ve decided to narrow our options to this 3:
+
+1. Lunr
+1. Google search
+1. Docsearch by Algolia
```
Comment (nate-double-u) [inline docs/search-best-practice.md:8]: I wonder if it would be better to use a bulleted list here -- we can imply preference via order without explicitly ranking them. We should also link to the sections below, may make this page more manageable as it grows.

```suggestion
* [Lunr](#lunr)
* [Google search](#programmable-search-engine-by-google)
* [Docsearch by Algolia](#docsearch-by-algolia)
```

File: docs/search-best-practice.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+# Pros And Cons
```
Comment (nate-double-u) [inline docs/search-best-practice.md:10]: I suggest removing this title, and bumping the others up one level on the hierarchy.

File: docs/search-best-practice.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+- Full power for customization
+- Search index is completely managed and hosted by the owner.
+
+
+### Cons
+- Can be difficult to configure and setup (If a team is already using hugo/docsy, this should be *very* easy to setup).
+- Requires javascript knowledge
+
+# When Is It Best To Use One Over Another?
+If you are looking to create a search capability for your open source project without having to depend on a 3rd party service, then you should consider using [Lunr](https://lunrjs.com/). You can take a look at [this](https://github.com/vitessio/website/pull/1119) pull request to see how we used Lunr to implement search in Vitess docs website.
```
Comment (nate-double-u) [inline docs/search-best-practice.md:64]: Let's add a link to the docsy process here too.

File: docs/search-best-practice.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+### Cons
+- Can be difficult to configure and setup (If a team is already using hugo/docsy, this should be *very* easy to setup).
+- Requires javascript knowledge
+
+# When Is It Best To Use One Over Another?
+If you are looking to create a search capability for your open source project without having to depend on a 3rd party service, then you should consider using [Lunr](https://lunrjs.com/). You can take a look at [this](https://github.com/vitessio/website/pull/1119) pull request to see how we used Lunr to implement search in Vitess docs website.  
+
+If you are looking to setup search without caring too much about cost, customization, and 3rd party dependency, then you should consider going for [Docsearch by Algolia](https://docsearch.algolia.com/).  
+
+If you are looking to create a search for documentation irrespective of where it’s hosted (separate site), then you should consider the [Programmable Search Engine (Google search)](https://developers.google.com/custom-search/docs/overview).
```
Comment (nate-double-u) [inline docs/search-best-practice.md:68]: I'm not sure I'd be convinced to use Google Search based on where it's hosted. Is Google still the best at providing relevant results?

File: docs/search-best-practice.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+
+### Cons
+- Can be difficult to configure and setup (If a team is already using hugo/docsy, this should be *very* easy to setup).
+- Requires javascript knowledge
+
+# When Is It Best To Use One Over Another?
+If you are looking to create a search capability for your open source project without having to depend on a 3rd party service, then you should consider using [Lunr](https://lunrjs.com/). You can take a look at [this](https://github.com/vitessio/website/pull/1119) pull request to see how we used Lunr to implement search in Vitess docs website.  
+
+If you are looking to setup search without caring too much about cost, customization, and 3rd party dependency, then you should consider going for [Docsearch by Algolia](https://docsearch.algolia.com/).
```
Comment (nate-double-u) [inline docs/search-best-practice.md:66]: I'm not sure we should suggest Algolia -- i'm happy to have it as an option listed above, but i'm not sure we have a use case for it as good as Lunr or Google.

File: docs/search-best-practice.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
+# Best Practices for Documentation Search
+Documentation (docs) search is an important part of developer experience. It makes life so easy for developers by enabling them narrow down to sections they’re looking for in the documentation. Choosing the right tool for docs search can be a hassle. 
+
+Here at CNCF,we’ve explored different search tools and we’ve decided to narrow our options to this 3:
```
Comment (nate-double-u) [inline docs/search-best-practice.md:4]: We can simplify the introduction here. Something like:

```suggestion
This page describes some common alternatives for static site search.
```

**Review by nate-double-u — COMMENTED** [review]
Thanks for the updates @thisisobate! I've made some more comments in line.

File: docs/search-best-practice.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+### Pros
+- Easy to configure and setup
+- Multi-language support
+- Support for image search
+- Search across a specified collection of sites or pages
+- No daily limits for query or records.
+
+
+### Cons
+- Google search [does not work in various countries](https://techcrunch.com/2010/04/19/google-censorship/).
```
[the remark this answers is not in the document]
Comment (nate-double-u) [reply docs/search-best-practice.md:41]: Here's another good link, tho still old: https://googleblog.blogspot.com/2010/04/controversial-content-and-free.html I wonder if they've published an update?

[the remark this answers is not in the document]
Comment (thisisobate) [reply docs/search-best-practice.md:41]: I found [this one](https://www.makeuseof.com/countries-google-doesnt-dominate/).

File: docs/search-best-practice.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+1. Docsearch by Algolia
+
+# Pros And Cons
+
+## Docsearch by Algolia
+[DocSearch](https://docsearch.algolia.com/) is a search tool powered by the Algolia search engine that crawls your docs and provides a dropdown search experience on your website.
+
+### Pros
+- Provides Front-end widgets out of the box: search input, dynamic positioning of search results, etc.
+- Easy to configure and setup
```
[the remark this answers is not in the document]
Comment (thisisobate) [reply docs/search-best-practice.md:17]: Yeah...maybe it's worth removing that line entirely.
Will fix up

File: docs/search-best-practice.md (hunk 9 above)
Comment (thisisobate) [reply docs/search-best-practice.md:66]: Alright.

File: docs/search-best-practice.md (hunk 10 above)
Comment (thisisobate) [reply docs/search-best-practice.md:4]: Beautiful...sounds a lot better 👍

File: docs/search-best-practice.md (hunk 8 above)
Comment (thisisobate) [reply docs/search-best-practice.md:68]: I think I used the wrong words here. Let me rephrase:

> If you are looking to create a search engine that not only focuses on the contents of one website (site search), but on a particular topic from multiple sites, then you should consider the [Programmable Search Engine (Google search)](https://developers.google.com/custom-search/docs/overview).

What do you think?

File: docs/searching-documentation.md [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+[DocSearch](https://docsearch.algolia.com/) is a search tool powered by the Algolia search engine that crawls your docs and provides a dropdown search experience on your website.
+
+## Pros
+- Provides Front-end widgets out of the box: search input, dynamic positioning of search results, etc.
+- Integrations with popular frameworks
+- Support for multi-language search.
+
+## Cons
+- The search index is wholly managed and hosted on Algolia servers
+- Hard to customize search
```
Comment (nate-double-u) [inline docs/searching-documentation.md:18]: let's remove this line for now.

File: docs/searching-documentation.md [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+- Support for image search
+- Search across a specified collection of sites or pages
+- No daily limits for queries or records.
+
+
+## Cons
+- Google search [does not work in various countries](https://www.makeuseof.com/countries-google-doesnt-dominate/).
+- Shows ads in between search results
+- Works only on deployed, production version of your site
+- Hard to customize
```
Comment (nate-double-u) [inline docs/searching-documentation.md:39]: Let's remove this line for now.

File: docs/searching-documentation.md [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+- Support for offline search
+- No external package dependency
+- Completely free and open source
+- Full power for customization
+- Support for multi-language search
+- Search index is completely managed and hosted by the owner.
+
+
+## Cons
+- Can be difficult to configure and setup (If a team is already using hugo/docsy, this should be *very* easy to setup).
```
Comment (nate-double-u) [inline docs/searching-documentation.md:56]: Subtle term shift suggestion:
```suggestion
- Can be complex to configure and setup (If a team is already using hugo/docsy, this should be *very* easy to setup).
```

**Review by nate-double-u — APPROVED** [review]
Let's make one more set of changes before we merge this in, otherwise it looks good to me.

If we aren't going to explain what makes something "hard" to configure or use, let's not go into it. We have suggestions made in order of preference, that should be enough for now. We can come back to this later and add some detail around how things get setup if needed.
